library/std/src/sys/uefi/mod.rs

```diff
@@ -0,0 +1,140 @@
+//! [`OsString`]), allows to set permissions more granularly, extract low-level
+//! file descriptors from files and sockets, and has platform-specific helpers
+//! for spawning processes.
+//!
+//! [`OsStr`]: crate::ffi::OsStr
+//! [`OsString`]: crate::ffi::OsString
+
+#![deny(unsafe_op_in_unsafe_fn)]
+pub mod alloc;
+pub mod args;
+#[path = "../unix/cmath.rs"]
+pub mod cmath;
+pub mod env;
+pub mod fs;
+pub mod io;
+#[path = "../unsupported/locks/mod.rs"]
+pub mod locks;
+pub mod net;
+pub mod os;
+#[path = "../windows/os_str.rs"]
+pub mod os_str;
+pub mod path;
+pub mod pipe;
+pub mod process;
+pub mod stdio;
+pub mod thread;
+#[path = "../unsupported/thread_local_key.rs"]
+pub mod thread_local_key;
+pub mod time;
+
+pub(crate) mod common;
+
+#[cfg(test)]
+mod tests;
+
+use crate::io as std_io;
+use crate::os::uefi;
+use crate::ptr::NonNull;
+
+pub mod memchr {
+    pub use core::slice::memchr::{memchr, memrchr};
+}
+
+pub unsafe fn init(argc: isize, argv: *const *const u8, _sigpipe: u8) {
+    assert_eq!(argc, 2);
+    let image_handle = unsafe { NonNull::new(*argv as *mut crate::ffi::c_void).unwrap() };
+    let system_table = unsafe { NonNull::new(*argv.add(1) as *mut crate::ffi::c_void).unwrap() };
+    unsafe { crate::os::uefi::env::init_globals(image_handle, system_table) };
+}
+
+// SAFETY: must be called only once during runtime cleanup.
+// NOTE: this is not guaranteed to run, for example when the program aborts.
+pub unsafe fn cleanup() {}
+
+#[inline]
+pub const fn unsupported<T>() -> std_io::Result<T> {
+    Err(unsupported_err())
+}
+
+#[inline]
+pub const fn unsupported_err() -> std_io::Error {
+    std_io::const_io_error!(
+        std_io::ErrorKind::Unsupported,
+        "operation not supported on this platform",
+    )
+}
+
+pub fn decode_error_kind(code: i32) -> crate::io::ErrorKind {
+    use crate::io::ErrorKind;
+    use r_efi::efi::Status;
+
+    if let Ok(code) = usize::try_from(code) {
+        common::status_to_io_error(Status::from_usize(code)).kind()
+    } else {
+        ErrorKind::Uncategorized
+    }
+}
+
+pub fn abort_internal() -> ! {
+    if let (Some(boot_services), Some(handle)) =
+        (common::try_boot_services(), uefi::env::try_image_handle())
+    {
+        let _ = unsafe {
+            ((*boot_services.as_ptr()).exit)(
+                handle.as_ptr(),
+                r_efi::efi::Status::ABORTED,
+                0,
+                crate::ptr::null_mut(),
+            )
+        };
+    }
+
+    // In case SystemTable and ImageHandle cannot be reached, use `core::intrinsics::abort`
+    core::intrinsics::abort();
+}
+
+// This function is needed by the panic runtime. The symbol is named in
+// pre-link args for the target specification, so keep that in sync.
+#[cfg(not(test))]
+#[no_mangle]
+pub extern "C" fn __rust_abort() {
+    abort_internal();
+}
+
+#[inline]
+pub fn hashmap_random_keys() -> (u64, u64) {
+    unsafe { (get_random().unwrap_or(1), get_random().unwrap_or(2)) }
```

Comment:
While `../unsupported` returns `(1, 2)` from `hashmap_random_keys()`, is that really the right choice here? My understanding is that the only reason `HashMap` and friends are not in `alloc` is that we don't want to provide them unless backed by real randomness.

Reply:
So should I `unwrap`? I can also use a PRNG with time as seed if that works better.

Reply:
> My understanding is that the only reason HashMap and friends are not in alloc is that we don't want to provide them unless backed by real randomness.

It is because liballoc can't depend on any OS existing and thus even when there is an OS it could only use a fixed seed.

> I can also use a PRNG with time as seed if that works better.

That doesn't make much sense to me. The randomness here is to prevent HashDOS attacks. Time is predictable enough that you can likely perform a HashDOS attack anyway. In any case I don't think defending against a HashDOS attack is really necessary for UEFI programs. It only affects availability, not integrity or confidentiality. And UEFI programs are generally not exposed to the internet, so an attacker capable of performing a HashDOS attack can likely already hurt availability by either turning off the computer (in case of physical access) or by wiping the disk (in case of root access). If you are implementing something like netboot, an attacker could likely perform a DOS (denial of service) by responding very slowly to the requests anyway. I'm sure there are cases where not using randomness would enable a remote HashDOS attack that they couldn't do if there was randomness, but I personally don't think it is worth panicking in those cases. Especially because that is in itself effectively a DOS.